IGEL_LENOVO_M600_SOUND_FIXES
	  Sound fix/quirk for Lenovo M600 device.

	  Source: IGEL
diff --git a/include/sound/jack.h b/include/sound/jack.h
index 715b95983188..4a204c764474 100644
--- a/include/sound/jack.h
@@ -75,6 +75,21 @@ struct snd_jack {
 	void (*private_free)(struct snd_jack *);
 };
 
+#if defined(CONFIG_IGEL_OWN_DEVICE_SOUND_FIXES) || defined(CONFIG_IGEL_LENOVO_M600_SOUND_FIXES)
+/* gottwald@igel.com moved here from sound/core/jack.c */
+
+struct snd_jack_kctl {
+	struct snd_kcontrol *kctl;
+	struct list_head list;  /* list of controls belong to the same jack */
+	unsigned int mask_bits; /* only masked status bits are reported via kctl */
+        struct snd_jack *jack;  /* pointer to struct snd_jack */
+	bool sw_inject_enable;  /* allow to inject plug event via debugfs */
+#ifdef CONFIG_SND_JACK_INJECTION_DEBUG
+	struct dentry *jack_debugfs_root; /* jack_kctl debugfs root */
+#endif
+};
+#endif
+
 #ifdef CONFIG_SND_JACK
 
 int snd_jack_new(struct snd_card *card, const char *id, int type,
diff --git a/sound/core/jack.c b/sound/core/jack.c
index 93e357a23f17..23e36d44f4ea 100644
--- a/sound/core/jack.c
@@ -15,6 +15,9 @@
 #include <sound/core.h>
 #include <sound/control.h>
 
+#if !defined(CONFIG_IGEL_OWN_DEVICE_SOUND_FIXES) && !defined(CONFIG_IGEL_LENOVO_M600_SOUND_FIXES)
+/* gottwald@igel.com moved to include/sound/jack.h if SOUND FIXES are choosen
+ * because struct is also needed by sound/pci/hda/patch_realtek.c */
 struct snd_jack_kctl {
 	struct snd_kcontrol *kctl;
 	struct list_head list;  /* list of controls belong to the same jack */
diff --git a/sound/hda/codecs/realtek/alc269.c b/sound/hda/codecs/realtek/alc269.c
index c18c18f9e652..62f7619061e9 100644
--- a/sound/hda/codecs/realtek/alc269.c
@@ -3786,6 +3923,9 @@ enum {
 	ALC225_FIXUP_DISABLE_MIC_VREF,
 	ALC225_FIXUP_DELL1_MIC_NO_PRESENCE,
 	ALC295_FIXUP_DISABLE_DAC3,
+#ifdef CONFIG_IGEL_LENOVO_M600_SOUND_FIXES
+	ALC233_FIXUP_LENOVO_M600,
+#endif
 	ALC285_FIXUP_SPEAKER2_TO_DAC1,
 	ALC285_FIXUP_ASUS_SPEAKER2_TO_DAC1,
 	ALC285_FIXUP_ASUS_HEADSET_MIC,
diff --git a/sound/hda/codecs/realtek/alc269.c b/sound/hda/codecs/realtek/alc269.c
index c18c18f9e652..62f7619061e9 100644
--- a/sound/hda/codecs/realtek/alc269.c
@@ -4030,6 +4180,39 @@ static void alc287_fixup_lenovo_yoga_book_9i(struct hda_codec *codec,
 	__snd_hda_apply_fixup(codec, id, action, 0);
 }
 
+#if defined(CONFIG_IGEL_OWN_DEVICE_SOUND_FIXES) || defined(CONFIG_IGEL_LENOVO_M600_SOUND_FIXES)
+static void alc_rename_jack(struct hda_codec *codec, unsigned int nid,
+			    const char *name, const char *new_name)
+{
+	struct hda_jack_tbl *j;
+	struct snd_jack_kctl *jack_kctl, *tmp_jack_kctl;
+
+	j = snd_hda_jack_tbl_get(codec, nid);
+	if (NULL != j && NULL != j->jack) {
+		list_for_each_entry_safe(jack_kctl,
+					 tmp_jack_kctl, &(j->jack->kctl_list), list)
+		{
+			if (!strcmp(jack_kctl->kctl->id.name, name)) {
+				strcpy(jack_kctl->kctl->id.name, new_name);
+			}
+		}
+	}
+}
+
+#ifdef CONFIG_IGEL_LENOVO_M600_SOUND_FIXES
+/* hamburg@igel.com */
+static void alc233_fixup_lenovo_m600(struct hda_codec *codec,
+				     const struct hda_fixup *fix, int action)
+{
+	if (HDA_FIXUP_ACT_BUILD == action) {
+		/* rename double */
+		alc_rename_jack(codec, 0x19,
+				"Mic Jack", "Headset Mic Jack");
+	}
+}
+#endif
+#endif
+
 static const struct hda_fixup alc269_fixups[] = {
 	[ALC269_FIXUP_GPIO2] = {
 		.type = HDA_FIXUP_FUNC,
diff --git a/sound/hda/codecs/realtek/alc269.c b/sound/hda/codecs/realtek/alc269.c
index c18c18f9e652..62f7619061e9 100644
--- a/sound/hda/codecs/realtek/alc269.c
@@ -4886,6 +5069,13 @@ static const struct hda_fixup alc269_fixups[] = {
 		.type = HDA_FIXUP_FUNC,
 		.v.func = alc295_fixup_disable_dac3,
 	},
+#ifdef CONFIG_IGEL_LENOVO_M600_SOUND_FIXES
+	/* hamburg@igel.com */
+	[ALC233_FIXUP_LENOVO_M600] = {
+		.type = HDA_FIXUP_FUNC,
+		.v.func = alc233_fixup_lenovo_m600
+	},
+#endif
 	[ALC285_FIXUP_SPEAKER2_TO_DAC1] = {
 		.type = HDA_FIXUP_FUNC,
 		.v.func = alc285_fixup_speaker2_to_dac1,
diff --git a/sound/hda/codecs/realtek/alc269.c b/sound/hda/codecs/realtek/alc269.c
index c18c18f9e652..62f7619061e9 100644
--- a/sound/hda/codecs/realtek/alc269.c
@@ -7733,6 +7966,9 @@ static const struct hda_model_fixup alc269_fixup_models[] = {
 	{.id = ALC255_FIXUP_DELL_SPK_NOISE, .name = "dell-spk-noise"},
 	{.id = ALC225_FIXUP_DELL1_MIC_NO_PRESENCE, .name = "alc225-dell1"},
 	{.id = ALC295_FIXUP_DISABLE_DAC3, .name = "alc295-disable-dac3"},
+#ifdef CONFIG_IGEL_LENOVO_M600_SOUND_FIXES
+	{.id = ALC233_FIXUP_LENOVO_M600, .name = "lenovo-m600"},
+#endif
 	{.id = ALC285_FIXUP_SPEAKER2_TO_DAC1, .name = "alc285-speaker2-to-dac1"},
 	{.id = ALC280_FIXUP_HP_HEADSET_MIC, .name = "alc280-hp-headset"},
 	{.id = ALC221_FIXUP_HP_FRONT_MIC, .name = "alc221-hp-mic"},
diff --git a/sound/hda/codecs/realtek/alc662.c b/sound/hda/codecs/realtek/alc662.c
index 5073165d1f3c..792d5129daca 100644
--- a/sound/hda/codecs/realtek/alc662.c
@@ -255,6 +255,137 @@ static void alc_fixup_headset_mode_alc668(struct hda_codec *codec,
 	alc_fixup_headset_mode(codec, fix, action);
 }
 
+#if defined(CONFIG_IGEL_OWN_DEVICE_SOUND_FIXES) || defined(CONFIG_IGEL_LENOVO_M600_SOUND_FIXES)
+static void alc_rename_ctl(struct hda_codec *codec, unsigned int nid,
+			   const char *name, const char *new_name)
+{
+	int i;
+	struct hda_nid_item *items = codec->mixers.list;
+
+	for (i = 0; i < codec->mixers.used; i++) {
+		if (items[i].nid == nid && !strcmp(items[i].kctl->id.name, name)) {
+			strcpy(items[i].kctl->id.name, new_name);
+			break;
+		}
+	}
+}
+
+static void alc_rename_jack(struct hda_codec *codec, unsigned int nid,
+			    const char *name, const char *new_name)
+{
+	struct hda_jack_tbl *j;
+	struct snd_jack_kctl *jack_kctl, *tmp_jack_kctl;
+
+	j = snd_hda_jack_tbl_get(codec, nid);
+	if (NULL != j && NULL != j->jack) {
+		list_for_each_entry_safe(jack_kctl,
+					 tmp_jack_kctl, &(j->jack->kctl_list), list)
+		{
+			if (!strcmp(jack_kctl->kctl->id.name, name)) {
+				strcpy(jack_kctl->kctl->id.name, new_name);
+			}
+		}
+	}
+}
+#endif
+
+#ifdef CONFIG_IGEL_OWN_DEVICE_SOUND_FIXES
+static void alc662_fixup_limit_mic_boost(struct hda_codec *codec,
+					 const struct hda_fixup *fix,
+					 int action)
+{
+	if (action != HDA_FIXUP_ACT_PROBE)
+		return;
+
+	/* The mic boosts on level 2 and 3 are too noisy
+	   on the internal mic input. Therefore limit the boost to 0 or 1. */
+	snd_hda_override_amp_caps(codec, 0x18, HDA_INPUT,
+				  (0x00 << AC_AMPCAP_OFFSET_SHIFT) |
+				  (0x01 << AC_AMPCAP_NUM_STEPS_SHIFT) |
+				  (0x2f << AC_AMPCAP_STEP_SIZE_SHIFT) |
+				  (0 << AC_AMPCAP_MUTE_SHIFT));
+
+	/* limit Mic Playback Volume to avoid noise in sound output */
+	snd_hda_override_amp_caps(codec, 0x0b, HDA_INPUT,
+				  (0x00 << AC_AMPCAP_OFFSET_SHIFT) |
+				  (0x14 << AC_AMPCAP_NUM_STEPS_SHIFT) |
+				  (0x02 << AC_AMPCAP_STEP_SIZE_SHIFT) |
+				  (0 << AC_AMPCAP_MUTE_SHIFT));
+}
+
+static void alc_fixup_igel_m340c_ctl(struct hda_codec *codec,
+				     const struct hda_fixup *fix, int action)
+{
+	struct snd_kcontrol *kctl;
+
+	if (action == HDA_FIXUP_ACT_BUILD) {
+		kctl = snd_hda_find_mixer_ctl(codec, "Speaker Playback Volume");
+		if (kctl) {
+			strcpy(kctl->id.name, "Alt Speaker Playback Volume");
+		}
+		kctl = snd_hda_find_mixer_ctl(codec, "Speaker Playback Switch");
+		if (kctl) {
+			strcpy(kctl->id.name, "Alt Speaker Playback Switch");
+		}
+
+		kctl = snd_hda_find_mixer_ctl(codec, "Headphone Playback Volume");
+		if (kctl) {
+			strcpy(kctl->id.name, "Speaker Playback Volume");
+		}
+		kctl = snd_hda_find_mixer_ctl(codec, "Headphone Playback Switch");
+		if (kctl) {
+			strcpy(kctl->id.name, "Speaker Playback Switch");
+		}
+	}
+}
+
+/* hamburg@igel.com: rename speaker control */
+static void alc_fixup_igel_ud9_ctl(struct hda_codec *codec,
+				   const struct hda_fixup *fix, int action)
+{
+	struct snd_kcontrol *kctl;
+
+	if (action == HDA_FIXUP_ACT_BUILD) {
+		/* ud9 with the new BIOS */
+		kctl = snd_hda_find_mixer_ctl(codec, "Front Playback Volume");
+		if (kctl) {
+			strcpy(kctl->id.name, "Speaker Playback Volume");
+		}
+		kctl = snd_hda_find_mixer_ctl(codec, "Front Playback Switch");
+		if (kctl) {
+			strcpy(kctl->id.name, "Speaker Playback Switch");
+		}
+
+		/* ud9 with the old BIOS */
+		kctl = snd_hda_find_mixer_ctl(codec, "Master Playback Volume");
+		if (kctl) {
+			strcpy(kctl->id.name, "PCM Playback Volume");
+		}
+		kctl = snd_hda_find_mixer_ctl(codec, "Master Playback Switch");
+		if (kctl) {
+			strcpy(kctl->id.name, "PCM Playback Switch");
+		}
+		alc_rename_ctl(codec, 0x09, "Capture Volume", "Alt Capture Volume");
+		alc_rename_ctl(codec, 0x09, "Capture Switch", "Alt Capture Switch");
+	}
+}
+
+/* hamburg@igel.com */
+static void alc_fixup_igel_h830(struct hda_codec *codec,
+				const struct hda_fixup *fix, int action)
+{
+	if (HDA_FIXUP_ACT_BUILD == action) {
+		/* rename mixer control element for the line out on the rear side */
+		alc_rename_ctl(codec, 0x03,
+				"PCM Playback Volume", "Headphone Playback Volume");
+		alc_rename_ctl(codec, 0x1b,
+				"PCM Playback Switch", "Headphone Playback Switch");
+		alc_rename_jack(codec, 0x1b,
+				"Line Out Jack", "Headphone Jack");
+	}
+}
+#endif
+
 enum {
 	ALC662_FIXUP_ASPIRE,
 	ALC662_FIXUP_LED_GPIO1,
